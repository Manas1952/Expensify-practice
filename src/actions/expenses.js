export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
})