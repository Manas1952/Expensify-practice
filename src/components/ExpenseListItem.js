import moment from 'moment'
import numeral from 'numeral'
import React from 'react'
import { NavLink } from 'react-router-dom'

const ExpenseListItem = ({ id, description, createdAt, amount, note }) => (
  <div>
      <NavLink to={`/edit/${id}`} key={id}>{description} {numeral(amount / 100).format('$0,0.00')} {moment(createdAt).format('MMMM Do, YYYY')} {note} </NavLink>
  </div>
)

export default ExpenseListItem