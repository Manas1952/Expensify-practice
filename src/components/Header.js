import React from 'react'
import { NavLink } from 'react-router-dom'

export const Header = () => (
  <header>
    <h1>Expensify</h1>
  <NavLink to="/" >Home</NavLink>
  <NavLink to="/create" >Add</NavLink>
  </header>
)

export default Header