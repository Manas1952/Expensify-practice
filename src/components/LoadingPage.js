import React from 'react'

const LoadingPage = () => (
  <img src='/images/loader.gif' />
)

export default LoadingPage