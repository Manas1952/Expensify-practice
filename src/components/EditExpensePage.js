import React from 'react'
import { connect } from 'react-redux'
import { editExpense, removeExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

export class EditExpensePage extends React.Component {
  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id })
    this.props.history.push('/')
  }
  onSubmit = (expense) => {
    this.props.history.push('/')
    this.props.editExpense(this.props.expense.id, expense)
  }
  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove} >Remove Expense</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id
    })
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense(id))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
  // react-router-dom@6.2.1 not passes {match, params, history} as props, so we use v5