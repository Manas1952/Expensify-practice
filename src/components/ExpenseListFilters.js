// import 'react-dates/initialize'
import React from 'react'
import { DateRangePicker } from 'react-dates'
import { connect } from 'react-redux'
import { setStartDate, setEndDate, setTextFilter, sortByAmount, sortByDate } from '../actions/filters'

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }
  onTextChange = (e) => {
    const text = e.target.value
    this.props.setTextFilter(text)
  }
  onSortChange = (e) => {
    if (e.target.value == 'date') {
      this.props.sortByDate()
    }
    else {
      this.props.sortByAmount()
    }
  }
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }
  render() {
    return (
      <div>
        <input type="text" placeholder='Search Expenses' value={this.props.filters.text} onChange={this.onTextChange} />
        <select name="" id="" value={this.props.filters.sortBy} onChange={this.onSortChange} >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDateId='startDateId'
          endDateId='endDateId'
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          onDatesChange={this.onDatesChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
          readOnly={true}
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}
const mapDispatchToProps = (dispatch) => {
  return {
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)