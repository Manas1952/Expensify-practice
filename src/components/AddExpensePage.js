import React from 'react'
import { v4 as uuid } from 'uuid'
import Header from './Header'
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'
import { connect } from 'react-redux'

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    const id = uuid()
    // this.props.dispatch(addExpense({ id, ...expense }))
    this.props.startAddExpense(expense)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense))  // we do this because previously we were accessing addExpense through props(i.e. this.props.dispatch(addExpense({ id, ...expense }))), and during testing test case doesn't know what the addExpense is. So a ccess it as dispatch
  }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage)