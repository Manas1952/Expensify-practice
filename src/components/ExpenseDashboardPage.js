import React from 'react'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesList from './ExpensesList'
import ExpensesSummary from './ExpensesSummary'
import Header from './Header'

const ExpenseDashBoardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpensesList />
  </div>
)

export default ExpenseDashBoardPage