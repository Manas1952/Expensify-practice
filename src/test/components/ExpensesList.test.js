import { shallow } from 'enzyme'
import React from 'react'
import { ExpensesList } from '../../components/ExpensesList'
import expenses from '../fixtures/expenses'

test('should render ExpensesList', () => {
  const wrapper = shallow(<ExpensesList expenses={expenses} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesList with empty expenses', () => {
  const wrapper = shallow(<ExpensesList expenses={[]} />)
  expect(wrapper).toMatchSnapshot()
})