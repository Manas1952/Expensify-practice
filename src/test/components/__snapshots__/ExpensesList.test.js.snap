// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesList 1`] = `
<div>
  <ExpenseListItem
    amount={123}
    createdAt={0}
    description="Testing_1"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={23456}
    createdAt={-345600000}
    description="Testing_2"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={3456}
    createdAt={345600000}
    description="Testing_3"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should render ExpensesList with empty expenses 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;
